### ISSUE TITLE
Improve rendering performance

### ISSUE THREAD

**Issue by Jermolene** [Jermolene/TiddlyWiki5#1864]
The TiddlyWiki 5 rendering process is still too slow in some situations. Using the Chrome profiler, I've identified and fixed some of the bottlenecks.

Background: I've been doing some work with a dataset of 1,200 tiddlers with 6,800 distinct tags. The test process was to enable performance instrumentation in `startup.js`, and import the dataset as a JSON file. I then switched to the "More" tab (to cause a refresh including the contents of the "Tags" tab, which is highly resource intensive to compute), and noted the overall main refresh time.

Using an empty 5.1.9, I recorded refresh times of approximately 25 seconds. With the improvements discussed before, the time fell to around 10 seconds.

The initial optimisations are:
- Refactored implementation of `$tw.utils.each` so that it can be optimised (previously, it was marked by Chrome as not being optimisable)
- More efficient implementation of the `tags` filter operator
- More efficient implementations of `parseWhitespace` and `skipWhitespace` in the parser
- Use caching in `wiki.getTiddlerData` so that the data doesn't need to be re-extracted if the tiddler hasn't changed

I intend to do more investigation.

**Comment by pmario** [comment]
Just did a short test with win8.1, FF 39. Seems to be the same for tiddlywiki.com 
Activating the More: Tags tab uses about 800ms without mods and 420 with mods. 

Same results with Chromium 40

**Comment by Jermolene** [comment]
Yay, speeding up by 50% is pretty good.

I think I'd be finding and fixing more problems if only developer tools weren't so cryptically frustrating to use. And buggy, in the case of Chrome.

**Comment by pmario** [comment]
>Yay, speeding up by 50% is pretty good.

Yes. I'll run some tests with a 4 core and 8 core mobile device, with the new modifications. I'll use TW.com as a reference. I'll post the results. 

5.1.8 with the 8 core is "kind of OK", but far away from native. The 4 core imo just sucks. So especially for the 4 core I think it will be much smoother now.

**Comment by pmario** [comment]
Jeremy, the JSON data, that you use for the tests. Are the programatically created, like my [20'000  tiddlers test code](https://groups.google.com/d/msg/tiddlywiki/E9DMg4ZTccw/P1qjS_-HOjAJ)

IMO we should use the same test data, so the results are comparable.

**Comment by elearnguy** [comment]
Nice.

On Sunday, July 5, 2015, Jeremy Ruston <notifications@github.com> wrote:

> The TiddlyWiki 5 rendering process is still too slow in some situations.
> Using the Chrome profiler, I've identified and fixed some of the
> bottlenecks.
>
> Background: I've been doing some work with a dataset of 1,200 tiddlers
> with 6,800 distinct tags. The test process was to enable performance
> instrumentation in startup.js, and import the dataset as a JSON file. I
> then switched to the "More" tab (to cause a refresh including the contents
> of the "Tags" tab, which is highly resource intensive to compute), and
> noted the overall main refresh time.
>
> Using an empty 5.1.9, I recorded refresh times of approximately 25
> seconds. With the improvements discussed before, the time fell to around 10
> seconds.
>
> The initial optimisations are:
>
>    - Refactored implementation of $tw.utils.each so that it can be
>    optimised (previously, it was marked by Chrome as not being optimisable)
>    - More efficient implementation of the tags filter operator
>    - More efficient implementations of parseWhitespace and skipWhitespace
>    in the parser
>    - Use caching in wiki.getTiddlerData so that the data doesn't need to
>    be re-extracted if the tiddler hasn't changed
>
> I intend to do more investigation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Jermolene/TiddlyWiki5/issues/1864>.
>


-- 
Sent from Gmail Mobile

**Comment by sukima** [comment]
Out of curiosity. When you mentioned a *optimized version of `each`* I thought about the possibility of an asynchronous version. Granted it wouldn't speed up the loop but it would allow the UI to update and be responsive as the loop executes.

Obviously you would have to add a loading indicator before rendering the full tag list and or let the tag show up one at a time. (Kind of a cool animation) But the win would be a responsive UI.

**Comment by Jermolene** [comment]
Hi @pmario 

> JSON data, that you use for the tests. Are the programatically created, like my 20'000 tiddlers test code

No, it's a spreadsheet I'm working with in my day job. I've exported it to CSV and written a tiny JS hack to convert it to JSON.

>  I'll run some tests with a 4 core and 8 core mobile device, with the new modifications. I'll use TW.com as a reference. I'll post the results.

Great, thanks, I'll be looking forward to it. I'd encourage anyone else who knows how to use the profiling tools to dig in and see if they can spot any hotspots.

@sukima async rendering is indeed an interesting idea. One could imagine creating a widget that asynchronously rendered its children.

**Comment by pmario** [comment]
I was thinking about the possibility to delegate computational expensive code to workers. ... but this also increases complexity quite a bit. ... So we should try to identify and get the "low hanging fruits" first.

**Comment by Jermolene** [comment]
A good candidate for an asynchronous operation is performing a static HTML export of all tiddlers; it takes tens of seconds for tw5.com.

**Comment by sukima** [comment]
> One could imagine creating a widget that asynchronously rendered its children.

Although a noble idea, I can not think of a generic way to do this. The concept of asynchronous rendering in JavaScript is that each computation is deferred via a `setTimeout`. In the case of a wrapped widget the only computational unit would be the full synchronous rendering of the children under it and it wouldn't buy you much (thanks to the way JS implements *pseudo-asynchronicity*).

The only way this would add value would be if every widget supported incremental rendering that could be managed in an asynchronous way. This would probably radically change / augment the current Widget API.

> A good candidate for an asynchronous operation is performing a static HTML export of all tiddlers; it takes tens of seconds for tw5.com.

Oh that a good one. I like asynchronous JS code challenges. Perhaps that's something I could look into this weekend. (My only want would be to have Promises available in TiddlyWiki; Will have to make due with callbacks).

**Comment by Jermolene** [comment]
@sukima I'm OK with using promises, particularly if existing async code is refactored to use promises.

**Comment by sukima** [comment]
@Jermolene I had thought about that the concern is adding a promise library has a few draw backs:

* Native Promises are new and not widely supported. It would narrow the field of supported browsers that TW is compatible with.
* A good promise library adds bloat. Worth it if you end up using them extensively. (However this could be the right way to go to support asynchronous widgets.)
* The official Polyfill is lacking in features to the Native Promises. Since even Native Promises are lacking in features compared to a Promise library it leave an empty feeling inside.
* Most micro-libs I've seen are cool in that they are small and clever but most don't follow the spec, have quirks, and/or are lacking in features.

In short I think that the use case for Promises needs a deep dive to hammer out the _actual_ needs so the right promise implementation can be chosen.

**Comment by Jermolene** [comment]
That sounds like a persuasive argument, @sukima!

**Comment by pmario** [comment]
>In short I think that the use case for Promises needs a deep dive to hammer out the actual needs so the right promise implementation can be chosen.

As you wrote most 3rd party libs add bloat and most convenience features they add, are very opinionated and not very generic. ... IMO the biggest advantage they add is, that they fix the browser quirks and that should be it. 

so +1 for a deeper dive. .. I'd like to see, a wrapper for our http utility functions to make them easier for plugin authors. And may be a library to simply move compute expensive code out to workers.

**Comment by twMat** [comment]
*Great* to see this is going on! @Jermolene , out of curiosity, would you generally say the most critical bottlenecks for performance have been *identified*?

Like @pmario , I've had the thought to ask external coders but I wouldn't even know what to ask. For this reason, but even more for our competent people in-house who just may not know which are the cricical bottlenecks, is there documentation defining (specifying?) "the problems" in more detailed fashion?

**Comment by sukima** [comment]
@pmario I like the idea of off loading computation to workers. However, one of the pluses for TiddlyWiki is it's ability to run on most browsers. Are web workers available to the target browsers that TiddlyWiki supports? if not how do you / should we polyfill web workers?

**Comment by Jermolene** [comment]
> Great to see this is going on! @Jermolene , out of curiosity, would you generally say the most critical bottlenecks for performance have been identified?

No, I think there's plenty of scope for further analysis. These changes are pretty localised tweaks to hotspots. The real opportunity is at a much higher level: for example, further optimising the refresh cycle.

I don't think I've fully mastered the available profiling tools, but they are the best way to highlight the bottlenecks.

@sukima @pmario I don't think we'll be able to use webworkers in the single file configuration. The last time I checked, web workers only work when served as a separate .JS file from an HTTPS url.

**Comment by twMat** [comment]
Well it is clear I have contributed yet another smile to the community!
Web workers of the world - unite!

**Comment by elearnguy** [comment]
Way cool.  Tell your story!

http://www.youtube.com/yt/creators/

**Comment by CodaCodr** [comment]
@Jermolene 
> I don't think we'll be able to use webworkers in the single file configuration. The last time I checked, web workers only work when served as a separate .JS file from an HTTPS url.

Firefox doesn't apply restrictions, Chrome hides them behind a switch ` --allow-file-access-from-files`. Not sure about IE/Edge et al.

See http://www.html5rocks.com/en/tutorials/workers/basics/

**Comment by pmario** [comment]
@Jermolene ... While performance is always a good thing to have, I'd suggest to close this issue. We have done a lot to increase the performance since 2015. eg: New indexes, optimizing internal structures, linked-lists instead of arrays ... and so on..

**Comment by Jermolene** [comment]
Thanks @pmario
